runtime/makepatches.py

```diff
@@ -0,0 +1,113 @@
+sys.path.append(os.path.dirname(os.path.realpath(__file__)))  # Workaround for python 3.6's obtuse import system.
+from filehandling.srgshandler import readsrgs
+from commands import Commands
+
+
+def main(conffile=None):
+    commands.checkforupdates()
+
+    decompile_side(0)
+    decompile_side(1)
+
+
+def decompile_side(side=0):
+    filename = None
+    if side == 0:
+        commands.logger.info("== Decompiling Client ==")
+        filename = "minecraft"
+    elif side == 1:
+        commands.logger.info("== Decompiling Server ==")
+        filename = "minecraft_server"
+    else:
+        commands.logger.info("!! Side is not provided !!")
+        return True
+
+    if commands.checkjars(side):
+        currenttime = time.time()
+        commands.logger.info('> Creating SRGS')
+        commands.createsrgs(side)
+        commands.logger.info('> Applying SpecialSource')
+        commands.applyss(side)
+        commands.logger.info('> Applying Exceptor')
+        commands.applyexceptor(side)
+        commands.logger.info('> Unzipping the jar')
+        commands.extractjar(side)
+        commands.logger.info('> Applying jadretro')
+        commands.applyjadretro(side)
+        commands.logger.info('> Decompiling...')
+        commands.applyjad(side)
+        commands.logger.info("> Applying existing patch files...")
+        shutil.copytree("src/" + filename, "src/" + filename + "_unpatched")
+        commands.applypatches(side)
+        commands.logger.info("> Done in %.2f seconds" % (time.time() - currenttime))
+        commands.logger.info("> Source located in src.")
+
+    return False
+
+
+def mainpatches():
+    commands.checkforupdates()
+
+    makepatches(0)
+    makepatches(1)
+
+
+def makepatches(side):
+    filename = None
+    if side == 0:
+        commands.logger.info("== Generating Patches for Client ==")
+        filename = "minecraft"
+    elif side == 1:
+        commands.logger.info("== Generating Patches for Server ==")
+        filename = "minecraft_server"
+    else:
+        commands.logger.info("!! Side is not provided !!")
+        return
+
+    currenttime = time.time()
+
+    commands.logger.info("> Generating new patch files...")
+    cmd = ["diff", "-r", "-U3", "--exclude=*~", filename + "_unpatched/net", filename + "/net"]
+    with open("src/" + filename + ".patch", "w") as patchfile:
+        subprocess.run(cmd, shell=True, stdout=patchfile,  stderr=subprocess.STDOUT, cwd="src/")
+
+    commands.logger.info("> Done in %.2f seconds" % (time.time() - currenttime))
+    commands.logger.info("> Patch files are located in src.")
+
+
+if __name__ == "__main__":
+    parser = OptionParser(version="MCP %s" % Commands.MCPVersion)
+    parser.add_option("-c", "--config", dest="config", help="additional configuration file")
+    (options, args) = parser.parse_args()
+    commands = Commands(options.config)
+    commands.logger.info("> This script REQUIRES that you have diff or diffutils installed and added to your path.")
```

Comment:
Why do you need that? Use ``diff.exe`` from cygwin.

Reply:
You can use any kind of diff, so long as it uses standard diff syntax.

Reply:
Could we just download a diff.exe if they are on Windows?